File: sdk/dart/lib/src/storage.dart
import 'package:http/http.dart' as http;

import 'http_client.dart';
import 'types.dart';

class StorageModule {
  final BakendHttpClient httpClient;

  StorageModule({required BakendHttpClient http}) : httpClient = http;

  Future<FileMetadata> upload(
    List<int> bytes, {
    String filename = 'file',
    String visibility = 'protected',
  }) async {
    final request = http.MultipartRequest(
      'POST',
      Uri.parse('${httpClient.baseUrl}/api/storage/upload'),
    );
    request.fields['visibility'] = visibility;
    request.files.add(
      http.MultipartFile.fromBytes(
        'file',
        bytes,
        filename: filename,
      ),
    );

    final body = await httpClient.request<Map<String, dynamic>>(
      '/api/storage/upload',
      method: 'POST',
      body: request,
    );
    return FileMetadata.fromJson(body);
  }

  Future<List<int>> download(String id) async {
    return httpClient.request<List<int>>(
      '/api/storage/$id',
      parse: (response) => response.bodyBytes,
    );
  }

  String getDownloadUrl(String id) => '${httpClient.baseUrl}/api/storage/$id';

  Future<void> delete(String id) async {
    await httpClient.request<void>(
      '/api/storage/$id',
      method: 'DELETE',
    );
  }
}

File: sdk/javascript/src/storage.ts
import type { FileMetadata, HttpClient, UploadOptions } from "./types.ts";

export class StorageModule {
  private readonly http: HttpClient;

  constructor(http: HttpClient) {
    this.http = http;
  }

  async upload(
    file: Blob | File | ArrayBuffer,
    options: UploadOptions = {},
  ): Promise<FileMetadata> {
    const formData = new FormData();
    const visibility = options.visibility ?? "protected";

    if (file instanceof ArrayBuffer) {
      const name = options.filename ?? "file";
      formData.append("file", new Blob([file]), name);
    } else if (file instanceof Blob) {
      const name = options.filename ?? (file instanceof File ? file.name : "file");
      formData.append("file", file, name);
    } else {
      formData.append("file", file);
    }

    formData.append("visibility", visibility);

    return this.http.request<FileMetadata>("/api/storage/upload", {
      method: "POST",
      body: formData,
    });
  }

  async download(id: string): Promise<Blob> {
    return this.http.request<Blob>(`/api/storage/${id}`);
  }

  getDownloadUrl(id: string): string {
    return `${this.http.getBaseUrl()}/api/storage/${id}`;
  }

  async delete(id: string): Promise<void> {
    await this.http.request<void>(`/api/storage/${id}`, {
      method: "DELETE",
    });
  }
}
